### DISCUSSION TITLE
Detected illegal task-local value binding when using swift-dependencies

### DISCUSSION THREAD

**Discussion by pepicrft** [pointfreeco/swift-dependencies#195] [General]
I was tinkering with swift-dependencies to understand better the swift-dependencies package to make an informed decision before introducing it in one of my projects, and I came across a "detected illegal task-local value binding" from `TaskLocal` when trying to run a test. It's likely because I need to use the package more effectively, but I'd love to understand why. The example that I was using is this one: [ParallelTesting.zip](https://github.com/pointfreeco/swift-dependencies/files/14599957/ParallelTesting.zip). All I'm doing is:

- Instantiating a `Utility` struct that has a dependency, `.identifier` injected using `@Dependency`
- Using `withDependencies` in the test to instantiate `Utility` with a different `.identifier` value.
- Obtain the value from an awaited task group and assert that the value is expected.

My goal with that test was to auto-generate a thousand copies of that test to verify if the state is well-scoped, which should be according to [this proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0311-task-locals.md), and that everything works fine in a world with structured and non-structured concurrency. However, I couldn't reach that assessment phase because TaskLocal yields a `detected illegal task-local value binding` error.

What am I doing wrong?

**Comment by mbrandonw** [comment]
Hi @pepicrft, this is simply a limitation of `TaskLocal`s when used in `withTaskGroup`. You can't access task locals directly in the closure of `withTaskGroup`, but you can access them in the closure of `group.addTask`. So you would need to do something like this:

```swift
let got = await withTaskGroup(of: String.self, returning: String.self) { group in
  group.addTask {
    utility.value()
  }
  return await group.first(where: { _ in true })!
}
```

I believe the error message that Swift prints out is quite self-explanatory:

```
error: task-local: detected illegal task-local value binding at Dependencies/Dependency.swift:148.
Task-local values must only be set in a structured-context, such as: around any (synchronous or asynchronous function invocation), around an 'async let' declaration, or around a 'with(Throwing)TaskGroup(...){ ... }' invocation. Notably, binding a task-local value is illegal *within the body* of a withTaskGroup invocation.

The following example is illegal:

    await withTaskGroup(...) { group in 
        await <task-local>.withValue(1234) {
            group.spawn { ... }
        }
    }

And should be replaced by, either: setting the value for the entire group:

    // bind task-local for all tasks spawned within the group
    await <task-local>.withValue(1234) {
        await withTaskGroup(...) { group in
            group.spawn { ... }
        }
    }

or, inside the specific task-group child task:

    // bind-task-local for only specific child-task
    await withTaskGroup(...) { group in
        group.spawn {
            await <task-local>.withValue(1234) {
                ... 
            }
        }

        group.spawn { ... }
    }
```

Is there anything in the above that is confusing?

**Reply by pepicrft** [reply]
> Hi @pepicrft, this is simply a limitation of TaskLocals when used in withTaskGroup. You can't access task locals directly in the closure of withTaskGroup, but you can access them in the closure of group.addTask. So you would need to do something like this:

Thanks a lot @mbrandonw 🙏🏼. That's very clear. Do you know by any chance if Apple is looking into eliminating that limitation?

> I believe the error message that Swift prints out is quite self-explanatory:

It definitively is. I happened to skim too quickly through the examples.
